### PR TITLE
feat: timestamp controller

### PR DESCRIPTION
**Pull request by eyevana**
## What I did

1. Moved all the timestamp calculations and state to a separate controller
2. Refactored pf-timestamp to use the controller

## Testing Instructions

1. Verify all the examples in the docs and demo are working as expected

### CODE REVIEW

File: elements/pf-timestamp/pf-timestamp.ts [hunk 1]
```diff
@@ -45,22 +46,16 @@ export class PfTimestamp extends LitElement {
...
   get time() {
-    const { hour12, customFormat, dateFormat: dateStyle, timeFormat: timeStyle, utc } = this;
-    const timeZone = utc ? 'UTC' : undefined;
-    const formatOptions = customFormat || { hour12, dateStyle, timeStyle, timeZone };
-    const formattedDate = this.#date.toLocaleString(this.locale, formatOptions);
-    return this.relative ? this.#getTimeRelative(this.#date) : `${formattedDate}${this.displaySuffix ? ` ${this.displaySuffix}` : ''}`;
+    return this.#controller.time([this.dateFormat, this.timeFormat, this.customFormat, this.displaySuffix, this.locale, this.relative, this.utc, this.hour12]);
   }
 
+  #controller = new TimestampController(this);
```
Comment (bennypowers) [inline elements/pf-timestamp/pf-timestamp.ts:57]:
```suggestion
  #timestamp = new TimestampController(this);
```

File: elements/pf-timestamp/pf-timestamp.ts [hunk 2]
```diff
@@ -45,22 +46,16 @@ export class PfTimestamp extends LitElement {
...
+    return this.#controller.isoString;
   }
 
   get time() {
-    const { hour12, customFormat, dateFormat: dateStyle, timeFormat: timeStyle, utc } = this;
-    const timeZone = utc ? 'UTC' : undefined;
-    const formatOptions = customFormat || { hour12, dateStyle, timeStyle, timeZone };
-    const formattedDate = this.#date.toLocaleString(this.locale, formatOptions);
-    return this.relative ? this.#getTimeRelative(this.#date) : `${formattedDate}${this.displaySuffix ? ` ${this.displaySuffix}` : ''}`;
+    return this.#controller.time([this.dateFormat, this.timeFormat, this.customFormat, this.displaySuffix, this.locale, this.relative, this.utc, this.hour12]);
```
Comment (bennypowers) [inline elements/pf-timestamp/pf-timestamp.ts:54]:
```suggestion
    return this.#controller.time;
```

**Review by bennypowers — CHANGES_REQUESTED** [review]
changes in the host's config props should forward those changes onto the timestamp controller

File: core/pfe-core/controllers/timestamp-controller.ts [hunk 3]
```diff
@@ -2,6 +2,10 @@ import type { ReactiveController, ReactiveElement } from 'lit';
 
 export type DateTimeFormat = 'full' | 'long' | 'medium' | 'short';
 
+const options = ['dateFormat', 'timeFormat', 'customFormat', 'displaySuffix', 'locale', 'relative', 'utc', 'hour12'] as const;
```
Comment (bennypowers) [inline core/pfe-core/controllers/timestamp-controller.ts:5]:
```suggestion
export interface Options {
  dateFormat: DateTimeFormat;
  // etc
}

const defaults = {
  dateFormat: 'short',
  timeFormat: 'short',
} // etc
```

let the constructor take a `Partial<Options>` in it's second parameter, and have setting one of those request an update. here's one way you could do that:

```js
constructor(host, options) {
  this.#options = {}
  for (const [name, value] of Object.entries(defaults)) {
    this.#options[name] = options[name] ?? value;
    Object.defineProperty(this, 'name', {
      get() { return this.#options[name]; },
      set(v) { 
        this.#options[name] = v;
        host.requestUpdate();
      },
    });
  }
```

File: elements/pf-timestamp/pf-timestamp.ts [hunk 4]
```diff
@@ -45,72 +47,34 @@ export class PfTimestamp extends LitElement {
...
+  #timestamp = new TimestampController(this);
+
+  willUpdate(changedProperties: Map<string, any>) {
     if (!this.displaySuffix && this.utc) {
       this.displaySuffix = 'UTC';
     }
+
+    for (const [prop] of changedProperties) {
+      if (this.#timestamp.isOption(prop)) {
+        // @ts-ignore
```
Comment (eyevana) [inline elements/pf-timestamp/pf-timestamp.ts:67]: @bennypowers could not get over this TS error for the life of me. Incorporated your feedback, although I made `options` public and defined the accessor methods on that property instead.

Comment (bennypowers) [reply elements/pf-timestamp/pf-timestamp.ts]: All good, Ivana.

I fixed up the typescript errors by taking a slightly different tac with the controller APIs. Some of the tests are still failing, could just be config defaults :shrug: take a look when you get a minute. 

Thanks for working on this

**Review by bennypowers — APPROVED** [review]

File: core/pfe-core/controllers/timestamp-controller.ts [hunk 5]
```diff
@@ -0,0 +1,132 @@
+import type { ReactiveController, ReactiveControllerHost } from 'lit';
+
+export type DateTimeFormat = 'full' | 'long' | 'medium' | 'short';
+
+export interface TimestampOptions {
+  dateFormat?: DateTimeFormat;
+  timeFormat?: DateTimeFormat;
+  customFormat?: Intl.DateTimeFormatOptions;
```
Comment (eyevana) [inline core/pfe-core/controllers/timestamp-controller.ts:8]: Nice!

**Comment by eyevana** [comment]
Thank you Benny! Going to spend some time today checking out your changes. I'm sure I'll have questions.

**Pull request merged**